standard/src/main/java/net/kyori/moonshine/strategy/StandardPlaceholderResolverStrategy.java

```diff
@@ -0,0 +1,168 @@
+import net.kyori.moonshine.placeholder.IPlaceholderResolver;
+import net.kyori.moonshine.strategy.supertype.ISupertypeStrategy;
+import net.kyori.moonshine.util.Either;
+import net.kyori.moonshine.util.Weighted;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+@ThreadSafe
+public final class StandardPlaceholderResolverStrategy<R, I, F> implements
+    IPlaceholderResolverStrategy<R, I, F> {
+  private final ISupertypeStrategy supertypeStrategy;
+
+  public StandardPlaceholderResolverStrategy(final ISupertypeStrategy supertypeStrategy) {
+    this.supertypeStrategy = supertypeStrategy;
+  }
+
+  @Override
+  public Map<String, ? extends F> resolvePlaceholders(final Moonshine<R, I, ?, F> moonshine,
+      final R receiver, final I intermediateText,
+      final MoonshineMethod<? extends R> moonshineMethod,
+      final @Nullable Object[] parameters)
+      throws PlaceholderResolvingException {
+    if (parameters.length == 0) {
+      return Collections.emptyMap();
+    }
+
+    final Map<String, F> finalisedPlaceholders = new LinkedHashMap<>(parameters.length);
+    final Map<String, ContinuanceValue<?>> resolvingPlaceholders = new LinkedHashMap<>(16);
+    final Parameter[] methodParameters = moonshineMethod.reflectMethod().getParameters();
+    final Type[] exactParameterTypes = GenericTypeReflector.getParameterTypes(
+        moonshineMethod.reflectMethod(), moonshine.proxiedType().getType());
+
+    for (int idx = 0; idx < parameters.length; ++idx) {
+      final Parameter parameter = methodParameters[idx];
+      final @Nullable Object value = parameters[idx];
+      if (value == null) {
+        // Nothing to resolve with.
+        continue;
+      }
+
+      final Type parameterType = GenericTypeReflector.getExactSubType(
+          exactParameterTypes[idx], value.getClass());
+
+      final @Nullable Placeholder placeholder = parameter.getAnnotation(Placeholder.class);
+      //noinspection ConstantConditions -- Javadocs say it returns null if absent.
+      if (placeholder == null) {
+        // Nothing to resolve.
+        continue;
+      }
+
+      final String placeholderName = placeholder.value().isEmpty()
+          ? parameter.getName()
+          : placeholder.value();
+      resolvingPlaceholders
+          .put(placeholderName, ContinuanceValue.continuanceValue(value, parameterType));
+    }
+
+    this.resolvePlaceholder(moonshine, receiver, finalisedPlaceholders,
+        resolvingPlaceholders, moonshineMethod, parameters);
+
+    return finalisedPlaceholders;
+  }
+
+  /**
+   * Resolve a single placeholder.
+   *
+   * @param moonshine             the moonshine instance
+   * @param finalisedPlaceholders the finalised placeholders
+   * @param resolvingPlaceholders the placeholders to resolve
+   * @param moonshineMethod       the method we are resolving a placeholder for
+   */
+  private void resolvePlaceholder(final Moonshine<R, I, ?, F> moonshine, final R receiver,
```

Comment:
Some black magic going on here. This could benefit from some refactoring or at least some comments explaining what is going on exactly